src/Service/AbstractService.php

```diff
@@ -11,26 +11,82 @@
 use ShipEngine\ShipEngineClient;
 
 /**
- * Serialize and send HTTP requests.
+ * Serialize and send RPC requests over HTTP messages.
  */
 abstract class AbstractService
 {
+    /**
+     * @var ShipEngineClient
+     */
     protected ShipEngineClient $client;
+
+    /**
+     * @var MessageFactory
+     */
     protected MessageFactory $message_factory;
 
+    /**
+     * @var string
+     */
+    private const JSON_RPC_SPEC = '2.0';
+
+    /**
+     * @var string
+     */
+    private const RPC_PATH = '/';
+
+    /**
+     * AbstractService constructor.
+     * @param ShipEngineClient $client
+     */
     public function __construct(ShipEngineClient $client)
     {
         $this->client = $client;
         $this->message_factory = MessageFactoryDiscovery::find();
     }
-    
+
     /**
-     * Create and send an HTTP request.
+     * Create and send an RPC request over HTTP messages..
+     *
+     * @param string $method name of an RPC method
+     * @param array $params data that a remote procedure will make use of
+     * @return ResponseInterface
      */
-    protected function request(string $method, string $path, string $body = null): ResponseInterface
+    protected function request(string $method, array $params): ResponseInterface
     {
-        $request = $this->message_factory->createRequest($method, $path, array(), $body);
+        $HTTP_METHOD = 'POST';
+
+        $jsonData = json_encode(array_filter([
+            'id' => $this->generateId(),
+            'jsonrpc' => self::JSON_RPC_SPEC,
+            'method' => $method,
+            'params' => $params
+        ]));
+
+        $request = $this->message_factory->createRequest($HTTP_METHOD, self::RPC_PATH, array(), $jsonData);
 
         return $this->client->sendRequest($request);
     }
+
+    //Using a variant of the third proposed solution for a cryptographically secure ID generator
+    // - https://stackoverflow.com/questions/2040240/php-function-to-generate-v4-uuid
+    /**
+     * Generate a cryptographically secure ID.
+     *
+     * @param null $data
+     * @return string
+     */
+    private function generateId($data = null): string
```

Comment:
I would remove this function and just use a library like https://github.com/ramsey/uuid to create a UUID v4.

```
use Ramsey\Uuid\Uuid;

$uuid = Uuid::uuid4();
```